Allow access to const edm::{Event,EventSetup} from device::{Event,EventSetup}

makortel opened the pull request:
#### PR description:

Giving access to `edm::{Event,EventSetup}` from `device::{Event,EventSetup}` allows interaction with helper code needing `edm::{Event,EventSetup}` from Alpaka EDModules.

Need came up in a work by @AdrianoDee 

#### PR validation:

Code compiles

makortel committed:
Allow access to const edm::{Event,EventSetup} from device::{Event,EventSetup}

To be able to interact with helper code needing
edm::{Event,EventSetup} from Alpaka EDModules.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-40849/34305

- This PR adds an extra 12KB to repository

cmsbuild commented:
A new Pull Request was created by  @makortel (Matti Kortelainen) for master.

It involves the following packages:

- HeterogeneousCore/AlpakaCore (**heterogeneous**)


@cmsbuild, @makortel, @fwyzard can you please review it and eventually sign? Thanks.
@missirol, @rovere this is something you requested to watch as well.
@perrotta, @dpiparo, @rappoccio you are the release manager for this.

cms-bot commands are listed <a href="http://cms-sw.github.io/cms-bot-cmssw-cmds.html">here</a>

makortel commented:
@cmsbuild, please test

fwyzard commented on HeterogeneousCore/AlpakaCore/interface/alpaka/Event.h:
```diff
@@ -47,6 +47,10 @@ namespace ALPAKA_ACCELERATOR_NAMESPACE::device {
 
+    // To be able to interact with non-Alpaka helper code that needs
+    // to access edm::Event
+    edm::Event const& edmEvent() const { return constEvent_; }
```
I'm fine with an explicit method like this.

So, just checking, what do you think about a conversion specifier like
```c++
operator edm::Event const&() const { return constEvent_; }
```
?

This would allow passing a `device::Event` transparently where an `edm::Event const&` is expected.

makortel replied:
I thought about this question and went back and forth, and in the end settled in favoring user convenience (which has been a theme for this version of the design), given that I could not think of a way how the implicit conversion could be misused or really caused unexpected behavior.

cmsbuild commented:
+1

**Summary**: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-4eaaa6/30824/summary.html
**COMMIT**: c37579102866f2844ab1b54c8e7a0ceabc2b3dcb
**CMSSW**: CMSSW_13_1_X_2023-02-22-2300/el8_amd64_gcc11
**User test area**: For local testing, you can use `/cvmfs/cms-ci.cern.ch/week1/cms-sw/cmssw/40849/30824/install.sh` to create a dev area with all the needed externals and cmssw changes.


## Comparison Summary



There are some workflows for which there are errors in the baseline:
 11634.596 step 3
 The results for the comparisons for these workflows could be incomplete 
This means most likely that the IB is having errors in the relvals.The error does NOT come from this pull request

**Summary**:
 - You potentially added 38 lines to the logs
 - Reco comparison results: 8 differences found in the comparisons
 - DQMHistoTests: Total files compared: 49
 - DQMHistoTests: Total histograms compared: 3528709
 - DQMHistoTests: Total failures: 3
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 3528684
 - DQMHistoTests: Total skipped: 22
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 48 files compared)
 - Checked 213 log files, 164 edm output root files, 49 DQM output files
 - [TriggerResults](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_13_1_X_2023-02-22-2300+4eaaa6/55673/triggerResults): no differences found

fwyzard commented:
skip it :-)

makortel commented:
Rebased on top of 5ebe8d8bac6 (tip of https://github.com/cms-sw/cmssw/pull/40725, and last commit of https://github.com/cms-sw/cmssw/pull/40832 before a merge commit from 13_1_X), and moved to use the implicit conversion operator.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-40849/34420

- This PR adds an extra 20KB to repository

cmsbuild commented:
Pull request #40849 was updated. @cmsbuild, @makortel, @fwyzard can you please check and sign again.

fwyzard commented:
please test

fwyzard commented:
+heterogeneous

The changes to the code and tests look good.

cmsbuild commented:
This pull request is fully signed and it will be integrated in one of the next master IBs after it passes the integration tests. This pull request will now be reviewed by the release team before it's merged. @perrotta, @dpiparo, @rappoccio (and backports should be raised in the release meeting by the corresponding L2)

fwyzard commented:
urgent

To be backported to 13.0.x for `CMSSW_13_0_0`.

cmsbuild commented:
+1

**Summary**: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-4eaaa6/31009/summary.html
**COMMIT**: abba24669fcc11ffa8ad41feb86237a5b859e073
**CMSSW**: CMSSW_13_1_X_2023-03-02-1100/el8_amd64_gcc11
**User test area**: For local testing, you can use `/cvmfs/cms-ci.cern.ch/week0/cms-sw/cmssw/40849/31009/install.sh` to create a dev area with all the needed externals and cmssw changes.


## Comparison Summary



**Summary**:
 - You potentially added 12 lines to the logs
 - Reco comparison results: 8 differences found in the comparisons
 - DQMHistoTests: Total files compared: 49
 - DQMHistoTests: Total histograms compared: 3529699
 - DQMHistoTests: Total failures: 8
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 3529669
 - DQMHistoTests: Total skipped: 22
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 48 files compared)
 - Checked 213 log files, 164 edm output root files, 49 DQM output files
 - [TriggerResults](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_13_1_X_2023-03-02-1100+4eaaa6/55806/triggerResults): no differences found

perrotta commented:
+1
